### ISSUE TITLE
DeepSeek V4 fused shared expert weights can be dropped in plugin path, causing severe accuracy loss

### ISSUE THREAD

**Issue by cubezhang** [ROCm/ATOM#1520]
## Summary

Fix DeepSeek V4 expert weight mapping so fused shared expert weights are included in `FusedMoE.make_expert_params_mapping()` when the owning MoE layer is built in fused-shared mode.

The previous lookup only detected modules whose class name was exactly `FusedMoE`. In the vLLM plugin path this can miss the actual fused MoE module/wrapper, causing `num_fused_shared` to stay `0`. The generic loader then rewrites `ffn.shared_experts.*` to `ffn.experts.{n_routed_experts}.*`, but `get_expert_mapping()` does not include that final shared slot, so the shared expert tensors are silently dropped and model quality collapses.

## Validation

- Patched service started successfully in a clean container on port `8002`.
- Loader log no longer reports `ffn.shared_experts.*` tensors being silently dropped.
- Completion smoke test: `3 cars * 4 wheels` returns `12`.
- `lm_eval --tasks gsm8k --num_fewshot 5 --limit 20`: `0.9` exact match.
- Prior full validation on the same fix: full GSM8K 5-shot recovered to `flexible-extract=0.9477`, `strict-match=0.9484`.

### Operating System

NAME="Ubuntu" VERSION="24.04.4 LTS (Noble Numbat)"

### CPU

AMD EPYC 9654 96-Core Processor

### GPU

GPU:   Name:                    AMD EPYC 9654 96-Core Processor       Marketing Name:          AMD EPYC 9654 96-Core Processor       Name:                    AMD EPYC 9654 96-Core Processor       Marketing Name:          AMD EPYC 9654 96-Core Processor       Name:                    gfx942                                Marketing Name:          AMD Instinct MI308X                       Name:                    amdgcn-amd-amdhsa--gfx942:sramecc+:xnack-       Name:                    amdgcn-amd-amdhsa--gfx9-4-generic:sramecc+:xnack-   Name:                    gfx942                                Marketing Name:          AMD Instinct MI308X                       Name:                    amdgcn-amd-amdhsa--gfx942:sramecc+:xnack-       Name:                    amdgcn-amd-amdhsa--gfx9-4-generic:sramecc+:xnack-   Name:                    gfx942                                Marketing Name:          AMD Instinct MI308X                       Name:                    amdgcn-amd-amdhsa--gfx942:sramecc+:xnack-       Name:                    amdgcn-amd-amdhsa--gfx9-4-generic:sramecc+:xnack-   Name:                    gfx942                                Marketing Name:          AMD Instinct MI308X                       Name:                    amdgcn-amd-amdhsa--gfx942:sramecc+:xnack-       Name:                    amdgcn-amd-amdhsa--gfx9-4-generic:sramecc+:xnack-   Name:                    gfx942                                Marketing Name:          AMD Instinct MI308X                       Name:                    amdgcn-amd-amdhsa--gfx942:sramecc+:xnack-       Name:                    amdgcn-amd-amdhsa--gfx9-4-generic:sramecc+:xnack-   Name:                    gfx942                                Marketing Name:          AMD Instinct MI308X                       Name:                    amdgcn-amd-amdhsa--gfx942:sramecc+:xnack-       Name:                    amdgcn-amd-amdhsa--gfx9-4-generic:sramecc+:xnack-   Name:                    gfx942                                Marketing Name:          AMD Instinct MI308X                       Name:                    amdgcn-amd-amdhsa--gfx942:sramecc+:xnack-       Name:                    amdgcn-amd-amdhsa--gfx9-4-generic:sramecc+:xnack-   Name:                    gfx942                                Marketing Name:          AMD Instinct MI308X                       Name:                    amdgcn-amd-amdhsa--gfx942:sramecc+:xnack-       Name:                    amdgcn-amd-amdhsa--gfx9-4-generic:sramecc+:xnack-

### ROCm Version

ROCm 7.2.4

### Steps to Reproduce

When serving `DeepSeek-V4-Flash-FP8` with ATOM/vLLM on ROCm, model quality collapsed because shared expert checkpoint tensors were silently dropped during weight loading.

The issue appears in `atom/models/deepseek_v4.py`, inside `DeepseekV4ForCausalLM.get_expert_mapping()`.

The current logic only detects fused shared experts by checking:

```python
if _m.__class__.__name__ == "FusedMoE":
In the vLLM plugin/wrapper path, this exact class-name check can miss the actual fused MoE module/wrapper. As a result, num_fused_shared stays 0, so get_expert_mapping() only creates mappings for routed experts and does not include the extra shared expert slot.
However, the generic ATOM loader still rewrites checkpoint keys like:
layers.N.ffn.shared_experts.w1.weight
layers.N.ffn.shared_experts.w2.weight
layers.N.ffn.shared_experts.w3.weight
to:
model.layers.N.ffn.experts.256.w1.weight
model.layers.N.ffn.experts.256.w2.weight
model.layers.N.ffn.experts.256.w3.weight
Because get_expert_mapping() did not include expert id 256, those tensors fail to map into experts.w13_* / experts.w2_* and get dropped.

Observed Log
The loader reported:
load_model: 258 checkpoint tensors were silently dropped because the rewritten name has no matching model parameter.

Examples:
layers.4.ffn.shared_experts.w1.weight -> model.layers.4.ffn.experts.w1.weight
layers.4.ffn.shared_experts.w2.weight -> model.layers.4.ffn.experts.w2.weight
layers.4.ffn.shared_experts.w3.weight -> model.layers.4.ffn.experts.w3.weight
This caused severe quality degradation.
Accuracy Before Fix
Full GSM8K 5-shot via lm_eval:
flexible-extract exact_match: 0.0083
strict-match     exact_match: 0.0000
Simple completion prompts also produced unstable / nonsensical outputs.
Proposed Fix
Change the fused shared expert detection in DeepseekV4ForCausalLM.get_expert_mapping().
Instead of relying only on exact class name FusedMoE, detect modules that expose num_fused_shared_experts. If that still returns 0, fall back to checking the owning MoE layer. If the MoE layer was constructed with _fuse_shared_into_routed=True, include self.args.n_shared_experts in the expert count.

Patch:
diff --git a/atom/models/deepseek_v4.py b/atom/models/deepseek_v4.py
index dada7adf..4350bf27 100644
--- a/atom/models/deepseek_v4.py
+++ b/atom/models/deepseek_v4.py
@@ -3054,9 +3054,21 @@ class DeepseekV4ForCausalLM(nn.Module):
         # mis-load expert weights -> garbage.
         num_fused_shared = 0
         for _m in self.model.modules():
-            if _m.__class__.__name__ == "FusedMoE":
+            if hasattr(_m, "num_fused_shared_experts"):
                 num_fused_shared = getattr(_m, "num_fused_shared_experts", 0)
                 break
+        if num_fused_shared == 0:
+            # Some plugin builds wrap/alias FusedMoE such that the exact class-name
+            # probe above misses it. If the owning MoE layer was constructed in
+            # fused-shared mode, the loader will rewrite ffn.shared_experts.* to
+            # ffn.experts.{n_routed_experts}.*; include that final slot here so the
+            # generic expert mapping loads it into w13/w2 instead of dropping it.
+            for _m in self.model.modules():
+                if _m.__class__.__name__ == "MoE" and getattr(
+                    _m, "_fuse_shared_into_routed", False
+                ):
+                    num_fused_shared = getattr(self.args, "n_shared_experts", 0)
+                    break
         num_experts = self.args.n_routed_experts + num_fused_shared
         return FusedMoE.make_expert_params_mapping(
             ckpt_gate_proj_name="w1",

Why This Fix Works

When fused shared experts are enabled, the shared expert occupies one extra expert slot after the routed experts.
For DeepSeek V4 Flash, routed experts are 0..255, and the fused shared expert should be mapped as expert id 256.
After this change:
num_experts = self.args.n_routed_experts + num_fused_shared
correctly becomes 257, so FusedMoE.make_expert_params_mapping() includes the shared expert slot. Then rewritten checkpoint keys such as ffn.experts.256.w1.* are loaded into the fused MoE parameters instead of being dropped.
Validation After Fix
After applying the patch in a clean container:
vLLM service started successfully.
Loader no longer reported ffn.shared_experts.* tensors being silently dropped.
Simple completion smoke test worked:Prompt: 3 cars * 4 wheels
Output included: 12

GSM8K small validation:
lm_eval --tasks gsm8k --num_fewshot 5 --limit 20
flexible-extract exact_match: 0.9
strict-match     exact_match: 0.9
Full GSM8K 5-shot validation recovered to:
flexible-extract exact_match: 0.9477 ± 0.0061
strict-match     exact_match: 0.9484 ± 0.0061

Environment
Model: DeepSeek-V4-Flash-FP8
Runtime: vLLM with ATOM plugin on ROCm
vLLM version observed: 0.22.1.dev0+g0b3ba88f1.d20260705
Container image: rocm/atom-dev:vllm-latest
Tensor parallel size: 4

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

<details>
<summary>rocminfo --support output</summary>

```
Paste output here
```

</details>


### Additional Information

_No response_

**Comment by yitingw1** [comment]
Related to https://github.com/ROCm/ATOM/pull/1527
